### PR TITLE
Evolve kubermatic-e2e to more general purpose tool

### PR DESCRIPTION
**Pull request by kron4eg**
**What this PR does / why we need it**:

**Release note**:
```release-note
NONE
```

### CODE REVIEW

**Comment by glower** [comment]
LGTM, will try it out today.

**Review by glower — DISMISSED** [review]
LGTM but let @mrIncompetent to have a look

File: api/cmd/kubermatic-e2e/cluster_creator.go
Comment (alvaroaleman) [inline api/cmd/kubermatic-e2e/cluster_creator.go:161]: How would in this scenario Ginkgo be called? Manually after a cluster was created?

Comment (kron4eg) [reply api/cmd/kubermatic-e2e/cluster_creator.go:161]: Yes manually, kubermatic-e2e will dump usercluster's admin kubeconfig, that subsequent tools can use to access created cluster.

EDIT:
manually in CI/CD scenario means steps :)

File: api/pkg/resources/cluster/cluster.go
Comment (p0lyn0mial) [inline api/pkg/resources/cluster/cluster.go:12]: what's the purpose of `resources` pkg in general ?

File: api/cmd/kubermatic-e2e/README.md
Comment (p0lyn0mial) [inline api/cmd/kubermatic-e2e/README.md]: should we change the name to `kubermatic-cluster-creator` ?

Comment (kron4eg) [reply api/cmd/kubermatic-e2e/README.md]: maybe, I'm open to suggestions

File: api/cmd/kubermatic-e2e/cluster_creator.go
Comment (alvaroaleman) [reply api/cmd/kubermatic-e2e/cluster_creator.go:161]: Yeah. Let's maybe have a chat on this next week or Friday, so @mrIncompetent can participate. Right now this breaks the way we run our e2e tests and does not update the docs, so in the current form it can not be merged

File: api/pkg/resources/cluster/cluster.go
Comment (kron4eg) [reply api/pkg/resources/cluster/cluster.go:12]: I'm not really sure, but apparently it's a warehouse for generated "resources" (configs, secrets, CRDs)

**Review by alvaroaleman — CHANGES_REQUESTED** [review]
At least the docs on how to run e2e tests must be updated

File: api/cmd/kubermatic-e2e/cluster_creator.go
Comment (kron4eg) [reply api/cmd/kubermatic-e2e/cluster_creator.go:161]: I had impression that kubermatic-e2e is not used anymore by out CI and was replaced by `api/cmd/conformance-tests` o_O

File: api/pkg/resources/cluster/cluster.go
Comment (p0lyn0mial) [reply api/pkg/resources/cluster/cluster.go:12]: Just thinking if it is okay if this pkg deals with api (external) types not crds (internal) :) ?

File: api/cmd/kubermatic-e2e/README.md
Comment (p0lyn0mial) [reply api/cmd/kubermatic-e2e/README.md]: will this tool run any tests ? if not then we could change the name if yes then we don't have to :)

File: api/pkg/resources/cluster/cluster.go
Comment (kron4eg) [reply api/pkg/resources/cluster/cluster.go:12]: there is similar package, but to deal with machines CRDs `api/pkg/resources/machine`, and it's used in kubermatic-e2e too.

File: api/cmd/kubermatic-e2e/cluster_creator.go
Comment (kron4eg) [inline api/cmd/kubermatic-e2e/cluster_creator.go]: @alvaroaleman BTW, here I've used `github.com/kubermatic/kubermatic/api/pkg/resources/machine.Machine()` directly, should it be replaced with `.MachineDeployment()` instead?

**Comment by kron4eg** [comment]
@alvaroaleman kubermatic-e2e is not used in our pipeline anymore

File: api/cmd/kubermatic-e2e/README.md
Comment (kron4eg) [reply api/cmd/kubermatic-e2e/README.md]: It used to run conformance tests, but not anymore. Conformance tests now run by the [new tool](https://github.com/kubermatic/kubermatic/tree/master/api/cmd/conformance-tests).
New purpose of this tool is to provide ability for end-users of kubermatic to create user-clusters  in their pipelines. But I don't have good names at the moment.

File: api/cmd/kubermatic-e2e/cluster_creator.go
Comment (alvaroaleman) [reply api/cmd/kubermatic-e2e/cluster_creator.go]: Yeah I think that makes sense

**Comment by alvaroaleman** [comment]
@kron4eg I see. Generally speaking I am not opposed to this but I would like to have an open discussion first as for what gurantees in terms of compatibility we provide and do not provider. Also, who gets to use this tool in the end.

BTW by now we should probably rename the tool since it doesn't do any testing anymore, maybe just `kubermatic-cli`?

**Comment by kron4eg** [comment]
> @kron4eg I see. Generally speaking I am not opposed to this but I would like to have an open discussion first as for what gurantees in terms of compatibility we provide and do not provider.

So, currently input specs are just HTTP models (which are supposedly stable). At this stage we don't promise anything, simply because we have to use internal cluster-api objects (even then are hidden behind resources/machines and resources/cluster facades). Once we will have Service Accounts, we will be able to fully utilize API — the we will continue to use same HTTP models but obviously — no direct cluster-api object, and THEN we will be able to promise something.

>Also, who gets to use this tool in the end.

One of our customers with pretty complex setup. For time being, it's enough for them to have what this tool provides. Later (with SAs in place) we will migrate them to actual `kubermatic-cli`.

> BTW by now we should probably rename the tool since it doesn't do any testing anymore, maybe just `kubermatic-cli`?

In absence of Service Accounts I don't think we can call it like this. `kubermatic-cli` should have subcommands for all those manipulations and listings that API allow, that's entirely different story and tool.

EDIT:
Consider this tool as "temporary" stub, that we have to use before we get proper service accounts. Once we have SAs, we should remove it entirely. But the thing is that we don't have SAs even planned.

**Comment by kron4eg** [comment]
Currently blocked by https://github.com/kubermatic/kubermatic/pull/2389

**Comment by kron4eg** [comment]
In order to maintain compatibility with 2.8 release, this PR will only do direct manipulations with cluster-api Machine objects.

After merge and cherry-pick to 2.8 release branch, there will be follow-up PR with cluster-api machineDeployment to master.

**Comment by alvaroaleman** [comment]
@kron4eg Please rename the tool to `kubermatic-cli`, other than that, LGTM :+1:

**Comment by kron4eg** [comment]
> @kron4eg Please rename the tool to `kubermatic-cli`, other than that, LGTM 👍

That name implies so much responsibility :D but OK.

File: api/cmd/kubermatic-cli/cluster_creator.go
Comment (alvaroaleman) [inline api/cmd/kubermatic-cli/cluster_creator.go:9]: Shouldn't we also move this to `kubermatic-cli`?

Comment (kron4eg) [reply api/cmd/kubermatic-cli/cluster_creator.go:9]: Sorry, I didn’t get it. What move? Where?

Comment (alvaroaleman) [reply api/cmd/kubermatic-cli/cluster_creator.go:9]: nvm, my github review plugin shows me the file `api/cmd/kubermatic-e2e/e2e_test_runner.go` as still existing but thats not the case, seems to be some bug with that plugin

**Comment by mrIncompetent** [comment]
Code looks good :)
It would have been great if the Makefile changes would be extracted into a dedicated PR, but i won't block this PR because of it.

I'm fine with pushing this out as a very early experimental tool - It would give us early feedback.
But we should avoid giving any guarantee on this CLI.

**Comment by mrIncompetent** [comment]
Please state in the Release note that this is an experimental CLI

**Comment by kron4eg** [comment]
@mrIncompetent done

**Review by mrIncompetent — APPROVED** [review]

**Review by alvaroaleman — APPROVED** [review]

**Review by p0lyn0mial — APPROVED** [review]

**Pull request merged**
